In [1]:
# 🚆 Delay Prediction using Machine Learning Regression

# Step 1: Import libraries

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor

from sklearn.metrics import mean_squared_error, r2_score


# Step 2: Load dataset

data = pd.read_csv(r"C:\Users\Nida\Downloads\railway_delay_dataset.csv")

print(data.head())


# Step 3: Convert categorical data to numeric

data = pd.get_dummies(data, drop_first=True)


# Step 4: Define features and target

X = data.drop("Delay_Minutes", axis=1)

y = data["Delay_Minutes"]


# Step 5: Split dataset

X_train, X_test, y_train, y_test = train_test_split(

    X, y, test_size=0.2, random_state=42

)


# -------------------------------------------------

# 🤖 MODEL 1: Linear Regression

print("----- Linear Regression -----")

lr = LinearRegression()

lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)

print("R2 Score:", r2_score(y_test, y_pred_lr))

print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_lr)))


# -------------------------------------------------

# 🌲 MODEL 2: Random Forest

print("\n----- Random Forest -----")

rf = RandomForestRegressor()

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

print("R2 Score:", r2_score(y_test, y_pred_rf))

print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_rf)))


# -------------------------------------------------

# ⚡ MODEL 3: AdaBoost

print("\n----- AdaBoost -----")

ab = AdaBoostRegressor()

ab.fit(X_train, y_train)

y_pred_ab = ab.predict(X_test)

print("R2 Score:", r2_score(y_test, y_pred_ab))

print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_ab)))


# -------------------------------------------------

# 🏆 Final Comparison

print("\nMODEL COMPARISON COMPLETE")

   Train_No    Station  Scheduled_Hour        Day Weather  Passenger_Count  \
0     10103        CST               6    Tuesday   Clear             1135   
1     10180    Byculla              21  Wednesday   Clear             1625   
2     10093        CST              18  Wednesday   Clear              382   
3     10015  Ghatkopar              17     Sunday   Clear             1888   
4     10107     Kalyan              13   Saturday    Rain              687   

   Delay_Minutes  Delayed  
0              2        0  
1              4        0  
2             22        1  
3              2        0  
4             12        1  
----- Linear Regression -----
R2 Score: 0.6953835344237278
RMSE: 4.174710123457029

----- Random Forest -----
R2 Score: 0.7424938650950124
RMSE: 3.8383398755191025

----- AdaBoost -----
R2 Score: 0.7377734255176114
RMSE: 3.873361116886158

MODEL COMPARISON COMPLETE


Three regression models were implemented to predict train delays: Linear Regression, Random Forest, and AdaBoost.
Random Forest achieved the highest R² score of 0.74 and lowest RMSE of 3.83, making it the best performing model for delay prediction.